Update to odML version 1.4 & add conversion from nix to odml

JuliaSprenger opened the pull request:
This assumes the odml to be saved as a metadata section with name 'odML document' in the NIX file as it is generated by the odML-to-nix converter.

The conversion is now depending on the input file format, so (.odML,. xml) is automatically converted to the corresponding .nix file and vice versa.

JuliaSprenger committed:
first steps for nix to odml, not working yet

JuliaSprenger committed:
fixed conversion from odml to nix for odml 1.4

JuliaSprenger committed:
add conversion from nix to odml

JuliaSprenger committed:
add comments, update readme

achilleas-k commented on convert.py:
```diff
@@ -81,44 +72,135 @@ def write_recurse(odmlseclist, nixparentsec):
+    nixsec.author = odmldoc.author
+    nixsec.date = odmldoc.date
+    nixsec.version = odmldoc.version
+    nixsec.repository = odmldoc.repository
```
Are these meant to be properties on the section? I don't see the attributes getting used anywhere later, so this probably has no effect, right?

achilleas-k commented on convert.py:
```diff
@@ -81,44 +72,135 @@ def write_recurse(odmlseclist, nixparentsec):
+
+def get_odml_doc(nix_file):
+    # identify odml document section in nix file
+    doc_section = nix_file.find_sections(lambda x: x.name=='odML document', limit=1)
```
Since the depth limit is 1, this search is a bit redundant and basically equivalent to:
```python
if "odML document" in nix_file.sections:
    doc_section = nix_file.sections["odML document"]
```

You can even skip the check further down and let NIXPy raise the `ValueError`, or you can catch it and rethrow with a custom message.

But that brings up another point: Is it a good idea to error out if the top-level document section isn't there? The metadata is still convertible. It would probably be more flexible to just convert everything and ignore any missing data, or just warn about it.

achilleas-k commented on convert.py:
```diff
@@ -81,44 +72,135 @@ def write_recurse(odmlseclist, nixparentsec):
+    doc_section = nix_file.find_sections(lambda x: x.name=='odML document', limit=1)
+    if not doc_section:
+        raise ValueError('No odML document section present in nix file.')
+    elif len(doc_section) > 1:
```
Given how the `find_sections` search above goes only 1 level deep and uses the section name to search, I don't think it's possible to get more than one section in the results.

Then again, this becomes unnecessary if we just access the section by name and catch or warn if it doesn't exist.

JuliaSprenger committed:
fix multiple conversion issues and add (incomplete) test

JuliaSprenger committed:
reformatting

JuliaSprenger commented:
@achilleas-k: I fixed the issues you mentioned above and added a simple test. However, I noticed that the back and forth conversion is a bit tricky so for now I am only comparing a selected number of attributes. Uncertainties are for example handled differently by nix and odml, so we need to find a way to guarantee the same dtype for these, too.

achilleas-k commented:
I'm going to merge and we can come back to the uncertainties when we finalise a common solution in NIX and odML.

Thanks for the PR.

I'll also open a couple of issues here and maybe set up CI at some point since we have tests.

JuliaSprenger commented:
I also suggest to change the repository name to something like 'odML-NIX-converter' since the current one might be misleading now.

achilleas-k commented:
Good point. Will do.
